File: apps/cargo-tracker/src/main/java/com/example/cargotracker/shared/domain/model/Location.java
package com.example.cargotracker.shared.domain.model;

import java.util.regex.Pattern;

/**
 * 地点。UN/LOCODE で識別する港・内陸地点。<strong>共有カーネル</strong>（ADR-005）。
 *
 * <p>共有カーネルに置いてよいのは本クラスと {@code ShipperId} の 2 要素のみである。
 * UN/LOCODE は国際標準であり、港の識別という意味はどの BC でも同一で解釈が分岐しない。
 *
 * @param unlocode UN/LOCODE（英大文字 5 文字。前 2 文字が国コード）
 */
public record Location(String unlocode) {

    private static final Pattern FORMAT = Pattern.compile("[A-Z]{2}[A-Z0-9]{3}");

    public Location {
        if (unlocode == null || !FORMAT.matcher(unlocode).matches()) {
            throw new IllegalArgumentException(
                    "地点は UN/LOCODE（英大文字 5 文字）で指定します: " + unlocode);
        }
    }

    public static Location of(String unlocode) {
        return new Location(unlocode);
    }
}

File: docs/article/source/csharp/apps/cargo-tracker/src/CargoTracker.Web/Shared/Domain/Model/Location.cs
namespace CargoTracker.Shared.Domain.Model;

/// <summary>
/// 位置情報（共有カーネル）。UN/LOCODE（国際規格の 5 文字コード）で港湾・地点を識別する。
/// 全コンテキストで共有する（domain-model・Shared Kernel）。
/// </summary>
public sealed record Location
{
    public string UnLocode { get; }
    public string Name { get; }

    public Location(string unLocode, string name = "")
    {
        if (string.IsNullOrWhiteSpace(unLocode) || unLocode.Length != 5)
        {
            throw new ArgumentException($"UN/LOCODE は 5 文字で指定してください: {unLocode}", nameof(unLocode));
        }
        UnLocode = unLocode.ToUpperInvariant();
        Name = name;
    }

    public bool SameAs(Location other) => UnLocode == other.UnLocode;
}
